src/backends/gitlab/API.js

```diff
@@ -0,0 +1,169 @@
+            ([key, val]) => `${ key }=${ encodeURIComponent(val) }`)
+          : [];
+    return `${ this.api_root }${ path }?${ [cacheBuster, ...encodedParams].join("&") }`;
+  }
+
+  request(path, options = {}) {
+    const headers = this.requestHeaders(options.headers || {});
+    const url = this.urlFor(path, options);
+    return fetch(url, { ...options, headers })
+    .then((response) => {
+      const contentType = response.headers.get("Content-Type");
+      if (options.method === "HEAD" || options.method === "DELETE") {
+        return Promise.all([response]);
+      }
+      if (contentType && contentType.match(/json/)) {
+        return Promise.all([response, response.json()]);
+      }
+      return Promise.all([response, response.text()]);
+    })
+    .catch(err => Promise.reject([err, null]))
+    .then(([response, value]) => (response.ok ? value : Promise.reject([value, response])))
+    .catch(([errorValue, response]) => {
+      const errorMessageProp = (errorValue && errorValue.message) ? errorValue.message : null;
+      const message = errorMessageProp || (isString(errorValue) ? errorValue : "");
+      throw new APIError(message, response && response.status, 'GitLab', { response, errorValue });
+    });
+  }
+  
+  readFile(path, sha, branch = this.branch) {
+    const cache = sha ? LocalForage.getItem(`gh.${ sha }`) : Promise.resolve(null);
+    return cache.then((cached) => {
+      if (cached) { return cached; }
+      
+      return this.request(`${ this.repoURL }/repository/files/${ encodeURIComponent(path) }/raw`, {
+        params: { ref: branch },
+        cache: "no-store",
+      })
+      .then((result) => {
+        if (sha) {
+          LocalForage.setItem(`gh.${ sha }`, result);
+        }
+        return result;
+      });
+    });
+  }
+
+  fileDownloadURL(path, branch = this.branch) {
+      return this.urlFor(`${ this.repoURL }/repository/files/${ encodeURIComponent(path) }/raw`, {
+        params: { ref: branch },
+      });
+  }
+  
+  fileExists(path, branch = this.branch) {
+    return this.request(`${ this.repoURL }/repository/files/${ encodeURIComponent(path) }`, {
+      method: "HEAD",
+      params: { ref: branch },
+      cache: "no-store",
+    }).then(() => true).catch(err => 
+      // 404 can mean either the file does not exist, or if an API
+      //   endpoint doesn't exist. We can't check this becaue we are
+      //   not getting the content with a HEAD request.
+      (err.status === 404 ? false : Promise.reject(err))
+    );
+  }
+
+  listFiles(path) {
+    return this.request(`${ this.repoURL }/repository/tree`, {
+      params: { path, ref: this.branch },
+    })
+    .then(files => files.filter(file => file.type === "blob"));
+  }
+
+  persistFiles(files, options) {
+    const uploads = files.map(async file => {
+      const exists = await this.fileExists(file.path);
+      return this.uploadAndCommit(file, {
+        commitMessage: options.commitMessage,
+        newFile: !exists,
```

Comment:
Most of this `newFile` logic can be removed, since the media library deletes old files before uploading new ones if they are being overwritten. The only place we need to keep it is for an entry, in which case we can check `options.newEntry`.